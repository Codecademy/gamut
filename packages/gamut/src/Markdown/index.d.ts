import { PureComponent } from 'react';
import { ManyOverrideSettings } from './libs/overrides';
export declare const preprocessingInstructions: {
    shouldPreprocessNode: (node: import("./libs/overrides").HTMLToReactNode) => boolean;
    preprocessNode: (node: import("./libs/overrides").HTMLToReactNode) => void;
}[];
export interface MarkdownProps {
    className?: string;
    inline?: boolean;
    overrides?: ManyOverrideSettings;
    spacing?: 'loose' | 'tight' | 'none';
    text?: string;
}
declare class Markdown extends PureComponent<MarkdownProps> {
    render(): JSX.Element;
}
export default Markdown;
