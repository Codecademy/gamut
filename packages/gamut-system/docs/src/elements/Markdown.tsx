import { MDXProvider } from '@mdx-js/react';
import MDXRenderer from 'gatsby-plugin-mdx/mdx-renderer';

import { Heading, Text } from './Text';
import { Link } from './Link';
import { List, ListItem } from './List';
import { Code, Highlighter } from './Code';
import { Table, Td, Th, Tr } from './Table';

const decorateProps = (props) => ({
  ...props,
  id: props.children.toLowerCase().split(' ').join('-'),
});

const components = {
  pre: Highlighter,
  a: ({ children, href }) => (
    <Link to={href.replace('/#', '#')}>{children}</Link>
  ),
  ul: List,
  li: ListItem,
  inlineCode: Code,
  table: Table,
  th: Th,
  td: Td,
  tr: Tr,
  p: (props) => <Text marginBottom={16}>{props.children}</Text>,
  h1: (props) => <Heading as="h1" {...decorateProps(props)} />,
  h2: (props) => <Heading as="h2" {...decorateProps(props)} />,
  h3: (props) => <Heading as="h3" {...decorateProps(props)} />,
  h4: (props) => <Heading as="h4" {...decorateProps(props)} />,
  h5: (props) => <Heading as="h5" {...decorateProps(props)} />,
  h6: (props) => <Heading as="h6" {...decorateProps(props)} />,
};

export const Markdown = ({ children }) => (
  <MDXProvider components={components}>
    <MDXRenderer scope={{}}>{children}</MDXRenderer>
  </MDXProvider>
);
