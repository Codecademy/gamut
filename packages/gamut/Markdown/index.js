import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import MarkdownJSX from 'markdown-to-jsx';
import s from './styles';

import Iframe from './overrides/Iframe';

const CODE_BLOCK_FENCED = /(`{3,}|~{3,}) *(\S+)? *\n([\s\S]+?)\s*\1 *(?:\n *)*/gim;

// Makes sure there is a leading newline on fenced code blocks
const ensureCodeBlockSpacing = str => str.replace(CODE_BLOCK_FENCED, '\n$&\n');

class Markdown extends PureComponent {
  static propTypes = {
    spacing: PropTypes.oneOf(['loose', 'tight', 'none']),
    overrides: PropTypes.object,
    className: PropTypes.string,
    inline: PropTypes.bool,
    text: PropTypes.string,
  };

  render() {
    const {
      spacing = 'tight',
      text = '',
      className,
      inline = false,
      overrides: userOverrides,
    } = this.props;

    const spacingStyles = s[`spacing-${spacing}`];
    const classes = cx(spacingStyles, className);

    const options = {
      overrides: {
        iframe: Iframe,
        ...userOverrides,
      },
      forceBlock: !inline,
      forceInline: inline,
    };

    return (
      <MarkdownJSX className={classes} options={options}>
        {ensureCodeBlockSpacing(text)}
      </MarkdownJSX>
    );
  }
}

export default Markdown;
