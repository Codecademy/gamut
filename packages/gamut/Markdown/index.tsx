import React, { PureComponent } from 'react';
import cx from 'classnames';
import marked from 'marked';
import HtmlToReact from 'html-to-react';
import insane from 'insane';
import omitProps from '../utils/omitProps';
import {
  createTagOverride,
  createCodeBlockOverride,
  ManyOverrideSettings,
} from './libs/overrides';
import s from './styles/index.scss';

import Iframe from './overrides/Iframe';
import Anchor from './overrides/Anchor';

const htmlToReactParser = new HtmlToReact.Parser({
  xmlMode: true,
});
const processNodeDefinitions = new HtmlToReact.ProcessNodeDefinitions();

const defaultSanitizationConfig = {
  allowedAttributes: {
    ...insane.defaults.allowedAttributes,
    span: ['class'],
    code: ['class'],
    pre: ['class'],
    source: ['src', 'type'],
    img: ['src', 'alt', 'height', 'width', 'title', 'aria-label', 'style'],
    video: ['width', 'height', 'align', 'style', 'controls'],
    iframe: [
      'src',
      'width',
      'height',
      'frameborder',
      'gesture',
      'allow',
      'allowfullscreen',
    ],
  },
  allowedClasses: {
    div: 'narrative-table-container',
  },
  allowedTags: [
    ...insane.defaults.allowedTags,
    'video',
    'source',
    'font',
    'pre',
    'code',
    'kbd',
    'br',
    'iframe',
    'codeblock',
  ],
};

const isValidNode = function() {
  return true;
};

export interface MarkdownProps {
  className?: string;
  inline?: boolean;
  overrides?: ManyOverrideSettings;
  spacing?: 'loose' | 'tight' | 'none';
  text?: string;
}

class Markdown extends PureComponent<MarkdownProps> {
  render() {
    const {
      spacing = 'tight',
      text = '',
      className,
      overrides: userOverrides = {},
      inline = false,
    } = this.props;

    if (!text) return null;

    const spacingStyles = s[`spacing-${spacing}`];
    const classes = cx(spacingStyles, className);

    const Wrapper = inline ? 'span' : 'div';

    const overrides = Object.keys(userOverrides).map(tagName => {
      if (tagName === 'CodeBlock') {
        return createCodeBlockOverride(tagName, userOverrides[tagName]);
      }
      return createTagOverride(tagName, userOverrides[tagName]);
    });

    const processingInstructions = [
      createTagOverride('iframe', {
        component: Iframe,
      }),
      createTagOverride('a', {
        component: Anchor,
      }),
      ...overrides,
      {
        shouldProcessNode() {
          return true;
        },
        processNode: processNodeDefinitions.processDefaultNode,
      },
    ];

    // Render markdown to html
    const rawHtml = inline ? marked.inlineLexer(text, []) : marked(text);
    const sanitizationConfig = {
      ...defaultSanitizationConfig,
      allowedTags: [
        ...defaultSanitizationConfig.allowedTags,
        ...Object.keys(userOverrides).map(tagName => tagName.toLowerCase()),
      ],
      allowedAttributes: {
        ...defaultSanitizationConfig.allowedAttributes,
        ...Object.keys(userOverrides).reduce((acc, tagName) => {
          return {
            ...acc,
            [tagName.toLowerCase()]: (
              userOverrides[tagName].allowedAttributes || []
            ).map(attr => attr.toLowerCase()),
          };
        }, {}),
      },
    };

    const html = insane(rawHtml, sanitizationConfig);

    // Render html to a react tree
    const react = htmlToReactParser.parseWithInstructions(
      html,
      isValidNode,
      processingInstructions
    );

    return (
      <Wrapper
        {...omitProps(Object.keys(this.props), this.props)}
        className={classes}
      >
        {react}
      </Wrapper>
    );
  }
}

export default Markdown;
