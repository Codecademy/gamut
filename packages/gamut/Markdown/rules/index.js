import SimpleMarkdown from 'simple-markdown';
import iframe from './iframe';

const rules = {
  ...SimpleMarkdown.defaultRules,
  iframe,
};

export default rules;
