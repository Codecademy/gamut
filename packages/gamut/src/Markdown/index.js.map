{"version":3,"file":"index.js","sourceRoot":"","sources":["index.tsx"],"names":[],"mappings":"AAAA,OAAO,KAAK,EAAE,EAAE,aAAa,EAAE,MAAM,OAAO,CAAC;AAC7C,OAAO,EAAE,MAAM,YAAY,CAAC;AAC5B,OAAO,MAAM,MAAM,QAAQ,CAAC;AAC5B,OAAO,MAAM,MAAM,QAAQ,CAAC;AAC5B,OAAO,WAAW,MAAM,eAAe,CAAC;AACxC,OAAO,SAAS,MAAM,oBAAoB,CAAC;AAC3C,OAAO,EACL,iBAAiB,EACjB,uBAAuB,EAEvB,iBAAiB,GAClB,MAAM,kBAAkB,CAAC;AAC1B,OAAO,yBAAyB,MAAM,2BAA2B,CAAC;AAClE,OAAO,EAAE,+BAA+B,EAAE,MAAM,sBAAsB,CAAC;AACvE,OAAO,CAAC,MAAM,qBAAqB,CAAC;AACpC,OAAO,MAAM,MAAM,yBAAyB,CAAC;AAC7C,OAAO,MAAM,MAAM,yBAAyB,CAAC;AAC7C,OAAO,KAAK,MAAM,wBAAwB,CAAC;AAE3C,MAAM,iBAAiB,GAAG,IAAI,WAAW,CAAC,MAAM,CAAC;IAC/C,OAAO,EAAE,IAAI;CACd,CAAC,CAAC;AAEH,MAAM,CAAC,MAAM,yBAAyB,GAAG,+BAA+B,CAAC,CAAC,CAAC,CAAC;AAE5E,MAAM,WAAW,GAAG;IAClB,OAAO,IAAI,CAAC;AACd,CAAC,CAAC;AAUF,MAAM,QAAS,SAAQ,aAA4B;IACjD,MAAM;QACJ,MAAM,EACJ,OAAO,GAAG,OAAO,EACjB,IAAI,GAAG,EAAE,EACT,SAAS,EACT,SAAS,EAAE,aAAa,GAAG,EAAE,EAC7B,MAAM,GAAG,KAAK,GACf,GAAG,IAAI,CAAC,KAAK,CAAC;QAEf,IAAI,CAAC,IAAI;YAAE,OAAO,IAAI,CAAC;QAEvB,MAAM,aAAa,GAAG,CAAC,CAAC,WAAW,OAAO,EAAE,CAAC,CAAC;QAC9C,MAAM,OAAO,GAAG,EAAE,CAAC,aAAa,EAAE,SAAS,CAAC,CAAC;QAE7C,MAAM,OAAO,GAAG,MAAM,CAAC,CAAC,CAAC,MAAM,CAAC,CAAC,CAAC,KAAK,CAAC;QAExC,MAAM,SAAS,GAAG,MAAM,CAAC,IAAI,CAAC,aAAa,CAAC,CAAC,GAAG,CAAC,OAAO,CAAC,EAAE;YACzD,IAAI,OAAO,KAAK,WAAW,EAAE;gBAC3B,OAAO,uBAAuB,CAAC,OAAO,EAAE,aAAa,CAAC,OAAO,CAAC,CAAC,CAAC;aACjE;YACD,OAAO,iBAAiB,CAAC,OAAO,EAAE,aAAa,CAAC,OAAO,CAAC,CAAC,CAAC;QAC5D,CAAC,CAAC,CAAC;QAEH,MAAM,sBAAsB,GAAG;YAC7B,iBAAiB,CAAC,QAAQ,EAAE;gBAC1B,SAAS,EAAE,MAAM;aAClB,CAAC;YACF,iBAAiB,CAAC,GAAG,EAAE;gBACrB,SAAS,EAAE,MAAM;aAClB,CAAC;YACF,iBAAiB,CAAC,OAAO,EAAE;gBACzB,SAAS,EAAE,KAAK,CAAC,EAAE,CAAC,CAClB,oBAAC,KAAK,kBAAC,SAAS,EAAE,OAAO,KAAK,OAAO,CAAC,CAAC,CAAC,GAAG,CAAC,CAAC,CAAC,GAAG,IAAM,KAAK,EAAI,CACjE;gBACD,iBAAiB,EAAE,CAAC,OAAO,CAAC;aAC7B,CAAC;YACF,GAAG,SAAS;YACZ,GAAG,iBAAiB;SACrB,CAAC;QAEF,MAAM,aAAa,GAAG;YACpB,WAAW,EAAE,IAAI;SAClB,CAAC;QAEF,0BAA0B;QAC1B,MAAM,OAAO,GAAG,MAAM;YACpB,CAAC,CAAC,MAAM,CAAC,WAAW,CAAC,IAAI,EAAE,EAAE,EAAE,aAAa,CAAC;YAC7C,CAAC,CAAC,MAAM,CAAC,IAAI,EAAE,aAAa,CAAC,CAAC;QAEhC,MAAM,kBAAkB,GAAG;YACzB,GAAG,yBAAyB;YAC5B,WAAW,EAAE;gBACX,GAAG,yBAAyB,CAAC,WAAW;gBACxC,GAAG,MAAM,CAAC,IAAI,CAAC,aAAa,CAAC,CAAC,GAAG,CAAC,OAAO,CAAC,EAAE,CAAC,OAAO,CAAC,WAAW,EAAE,CAAC;aACpE;YACD,iBAAiB,EAAE;gBACjB,GAAG,yBAAyB,CAAC,iBAAiB;gBAC9C,GAAG,MAAM,CAAC,IAAI,CAAC,aAAa,CAAC,CAAC,MAAM,CAAC,CAAC,GAAG,EAAE,OAAO,EAAE,EAAE;oBACpD,OAAO;wBACL,GAAG,GAAG;wBACN,CAAC,OAAO,CAAC,WAAW,EAAE,CAAC,EAAE,CACvB,aAAa,CAAC,OAAO,CAAC,CAAC,iBAAiB,IAAI,EAAE,CAC/C,CAAC,GAAG,CAAC,IAAI,CAAC,EAAE,CAAC,IAAI,CAAC,WAAW,EAAE,CAAC;qBAClC,CAAC;gBACJ,CAAC,EAAE,EAAE,CAAC;aACP;SACF,CAAC;QAEF,MAAM,IAAI,GAAG,MAAM,CAAC,OAAO,EAAE,kBAAkB,CAAC,CAAC;QAEjD,8BAA8B;QAC9B,MAAM,KAAK,GAAG,iBAAiB,CAAC,qBAAqB,CACnD,IAAI,EACJ,WAAW,EACX,sBAAsB,EACtB,yBAAyB,CAC1B,CAAC;QAEF,OAAO,CACL,oBAAC,OAAO,oBACF,SAAS,CAAC,MAAM,CAAC,IAAI,CAAC,IAAI,CAAC,KAAK,CAAC,EAAE,IAAI,CAAC,KAAK,CAAC,IAClD,SAAS,EAAE,OAAO,KAEjB,KAAK,CACE,CACX,CAAC;IACJ,CAAC;CACF;AAED,eAAe,QAAQ,CAAC"}