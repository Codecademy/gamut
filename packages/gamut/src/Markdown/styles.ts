import { css, Theme } from '@emotion/react';

// Base markdown styles mixin equivalent
const markdownStyle = (
  theme: Theme,
  fontSize = '1.1rem',
  lineHeight = 1.6,
  baseMargin = '1rem'
) => css`
  font-size: ${fontSize};

  .g-md-h1,
  .g-md-h2,
  .g-md-h3,
  .g-md-h4 {
    margin-top: 1.5rem;
    margin-bottom: 1.5rem;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
  }

  .g-md-p {
    line-height: ${lineHeight};
    word-wrap: break-word;
    overflow-wrap: break-word;
    word-break: break-word;

    &:empty {
      display: none;
    }

    &:last-child {
      margin-bottom: 0;
    }
  }

  .g-md-ul,
  .g-md-ol {
    list-style: none;
    padding-left: 0;
    margin: ${baseMargin} 0;

    .g-md-ul,
    .g-md-ol {
      /* Even out nested lists */
      margin-top: calc(${theme.spacing[16]} / 2);

      li:last-child {
        margin-bottom: 0;
      }
    }

    .g-md-li {
      position: relative;
      margin-left: ${theme.spacing[32]};
      margin-bottom: calc(${theme.spacing[16]} / 2);
      line-height: ${lineHeight};

      &:last-child {
        margin-bottom: ${theme.spacing[16]};
      }
    }
  }

  .g-md-ul .g-md-li::before {
    background: currentColor;
    border-radius: 100%;
    content: ' ';
    display: inline-block;
    height: 0.5rem;
    left: -1.25rem;
    line-height: 1.6;
    position: absolute;
    text-align: center;
    top: calc(1.6rem - ${theme.spacing[16]});
    width: 0.5rem;
  }

  .g-md-ol {
    counter-reset: 'ordered-list';

    > .g-md-li {
      counter-increment: ordered-list;

      &::before {
        background: transparent;
        content: counter(ordered-list) '.';
        position: absolute;
        top: 0;
        left: calc(-${theme.spacing[32]});
        width: ${theme.spacing[32]};
        text-align: center;
        line-height: inherit;
      }
    }
  }

  /* Inline code style */
  .g-md-code {
    padding: 0 0.25rem;
    margin: 0 0.0625rem;
    border-radius: 0.125rem;
    color: ${theme.colors['navy-900']};
    background-color: ${theme.colors['gray-200']};
    white-space: normal;
    vertical-align: baseline;
    font-family: ${theme.fontFamily.monospace};
    font-size: 0.8em;
  }

  .g-md-pre {
    white-space: pre-wrap;

    > .g-md-code {
      display: block;
      white-space: pre-wrap;
      padding: 0;
      margin: 0;
      color: inherit;
      background-color: transparent;
      font-family: ${theme.fontFamily.monospace};
    }
  }

  .g-md-img {
    display: flex;
    max-width: 100%;
    margin: ${baseMargin} auto 0;
  }

  .g-md-p .g-md-img {
    margin: 0 auto;
  }

  .g-md-iframe {
    width: 100%;
  }

  .g-md-blockquote {
    display: block;
    font-family: ${theme.fontFamily.base};
    padding-left: ${baseMargin};
    border-left: calc(${baseMargin} / 2) solid ${theme.colors['gray-200']};
    margin: 0;
    margin-bottom: ${baseMargin};
    font-style: italic;

    /* The Markdown processor combines multiple lines
     * into a single tag if the same tag repeats line to line
     * this is mainly for <p> tags since that's the default */
    > * {
      white-space: pre-wrap;
    }
  }

  .g-md-details {
    margin-bottom: ${baseMargin};

    .g-md-summary {
      font-weight: bold;
      cursor: pointer;
    }
  }

  .g-md-details[open] .g-md-summary {
    margin-bottom: ${baseMargin};
  }
`;

// Spacing styles
export const spacingLoose = (theme: Theme) => css`
  ${markdownStyle(theme, '1.2rem', 1.7)}
`;

export const spacingTight = (theme: Theme) => css`
  ${markdownStyle(theme)}
`;

export const spacingNone = () => css`
  /* No spacing styles applied */
`;

// Main styles object for easy access
export const markdownStyles = {
  loose: spacingLoose,
  tight: spacingTight,
  none: spacingNone,
};
