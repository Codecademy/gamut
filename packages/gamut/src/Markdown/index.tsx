import cx from 'classnames';
import HtmlToReact from 'html-to-react';
import marked from 'marked';
import React, { PureComponent } from 'react';
import sanitizeMarkdown from 'sanitize-markdown';

import { omitProps } from '../utils/omitProps';
import {
  createCodeBlockOverride,
  createTagOverride,
  ManyOverrideSettings,
  standardOverrides,
} from './libs/overrides';
import { Iframe } from './libs/overrides/Iframe';
import {
  MarkdownAnchor,
  MarkdownAnchorProps,
} from './libs/overrides/MarkdownAnchor';
import { Table, TableProps } from './libs/overrides/Table';
import { createPreprocessingInstructions } from './libs/preprocessing';
import { defaultSanitizationConfig } from './libs/sanitizationConfig';
import styles from './styles/index.module.scss';

const htmlToReactParser = new HtmlToReact.Parser({
  xmlMode: true,
});

const preprocessingInstructions = createPreprocessingInstructions(styles);

const isValidNode = () => true;

export type SkipDefaultOverridesSettings = {
  a?: boolean;
  iframe?: boolean;
  table?: boolean;
};

export type MarkdownProps = {
  className?: string;
  inline?: boolean;
  overrides?: ManyOverrideSettings;
  skipDefaultOverrides?: SkipDefaultOverridesSettings;
  /**
   * Enables generated header ids for H1-6 tags
   * Can generate duplicate IDs if used on separate markdown components on the same page
   */
  headerIds?: boolean;
  spacing?: 'loose' | 'tight' | 'none';
  text?: string;
  /**
   * Callback when a markdown anchor tag is clicked
   */
  onAnchorClick?: (event: React.MouseEvent<HTMLAnchorElement>) => void;
};

export class Markdown extends PureComponent<MarkdownProps> {
  render() {
    const {
      spacing = 'tight',
      text = '',
      className,
      overrides: userOverrides = {},
      skipDefaultOverrides = {},
      inline = false,
      headerIds,
      onAnchorClick,
    } = this.props;

    if (!text) return null;

    const spacingStyles = styles[`spacing-${spacing}`];
    const classes = cx(spacingStyles, className);

    const Wrapper = inline ? 'span' : 'div';

    const overrides = Object.keys(userOverrides).map((tagName) => {
      if (tagName === 'CodeBlock') {
        return createCodeBlockOverride(tagName, userOverrides[tagName]);
      }
      return createTagOverride(tagName, userOverrides[tagName]);
    });

    const processingInstructions = [
      !skipDefaultOverrides.iframe &&
        createTagOverride('iframe', {
          component: Iframe,
        }),
      !skipDefaultOverrides.a &&
        createTagOverride('a', {
          component: MarkdownAnchor,
          processNode: (node, props) => {
            // Note: this processNode override is necessary because wrapping this component
            // in an anonymous functional component as with the Table below causes react rendering
            // to crash with some chrome translation features.
            // See https://codecademy.atlassian.net/browse/WEB-1214
            return (
              <MarkdownAnchor
                onClick={onAnchorClick}
                {...(props as MarkdownAnchorProps)}
              />
            );
          },
        }),
      !skipDefaultOverrides.table &&
        createTagOverride('table', {
          component: (props: TableProps) => (
            <Table maxHeight={spacing === 'tight' ? 180 : 500} {...props} />
          ),
          allowedAttributes: ['style'],
        }),
      ...overrides,
      ...standardOverrides,
    ].filter(Boolean);

    const markedOptions = {
      smartypants: true,
      headerIds,
      headerPrefix: 'heading-',
    };

    // Render markdown to html
    const rawHtml = inline
      ? marked.inlineLexer(text, [], markedOptions)
      : marked(text, markedOptions);

    const sanitizationConfig = {
      ...defaultSanitizationConfig,
      allowedTags: [
        ...defaultSanitizationConfig.allowedTags,
        ...Object.keys(userOverrides).map((tagName) => tagName.toLowerCase()),
      ],
      allowedAttributes: {
        ...defaultSanitizationConfig.allowedAttributes,
        ...Object.keys(userOverrides).reduce((acc, tagName) => {
          return {
            ...acc,
            [tagName.toLowerCase()]: (
              userOverrides[tagName].allowedAttributes || []
            ).map((attr) => attr.toLowerCase()),
          };
        }, {}),
      },
    };

    const html = sanitizeMarkdown(rawHtml, sanitizationConfig);

    // Render html to a react tree
    const react = htmlToReactParser.parseWithInstructions(
      html,
      isValidNode,
      processingInstructions,
      preprocessingInstructions
    );

    return (
      <Wrapper
        {...omitProps(Object.keys(this.props), this.props)}
        className={classes}
      >
        {react}
      </Wrapper>
    );
  }
}
