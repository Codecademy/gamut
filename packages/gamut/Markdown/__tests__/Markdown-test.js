import React from 'react';
import { mount, render } from 'enzyme';
import Markdown from '../index';

const basicMarkdown = `
# Heading 1

### Heading 3

\`\`\`js
const jsCode = () => {
  console.log('hello world);
}
\`\`\`

`;

const htmlMarkdown = `
<h1>Heading 1</h1>

<h3>Heading 3</h3>

<iframe src="https://www.youtube.com/embed/KvgrQIK1yPY" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
`;

const youtubeMarkdown = `
<iframe src="https://www.youtube.com/embed/KvgrQIK1yPY" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
`;

const codeBlockBugMarkdown = `
having this text above the codeblock breaks the codeblock in markdown-in-jsx
\`\`\`js
const jsCode = () => {
  console.log('hello world);
}
\`\`\`

\`\`\`js
const jsCode = () => {
  console.log('hello world);
}
\`\`\`

`;

// test for working MIXED_INLINE_HTML_TAGS regex
const htmlWrappedMarkdown = `
Starting some html in the middle of a <div class="test">

# Heading one
</div>

`;

describe('<Markdown />', () => {
  it('renders standard Markdown', () => {
    const markdown = mount(<Markdown text={basicMarkdown} />);
    expect(markdown.find('h1').length).toEqual(1);
    expect(markdown.find('h3').length).toEqual(1);
    expect(markdown.find('code').length).toEqual(1);
  });

  it('Renders html content in markdown', () => {
    const markdown = mount(<Markdown text={htmlMarkdown} />);
    expect(markdown.find('h1').length).toEqual(1);
    expect(markdown.find('h3').length).toEqual(1);
    expect(markdown.find('iframe').length).toEqual(1);
  });

  it('Wraps youtube iframes in a flexible container', () => {
    const markdown = mount(<Markdown text={youtubeMarkdown} />);
    expect(markdown.find('[data-testid="yt-iframe"]').length).toEqual(1);
  });

  it('Wraps the markdown in a div by default (block)', () => {
    const markdown = mount(<Markdown text={basicMarkdown} />);
    expect(markdown.find('div.spacing-tight').length).toEqual(1);
  });

  it('Wraps the markdown in a span when inline', () => {
    const markdown = mount(<Markdown text={basicMarkdown} inline />);
    expect(markdown.find('span.spacing-tight').length).toEqual(1);
  });

  it('Fixes the newline bug with codeBlocks', () => {
    const markdown = mount(<Markdown text={codeBlockBugMarkdown} />);
    expect(markdown.find('code.lang-js').length).toEqual(2);
  });

  it('Allows passing in arbitrary react component overrides', () => {
    const TestComponent = () => <marquee>coooool</marquee>;

    const text = `
# Heading

<TestComponent />
    `;

    const overrides = {
      TestComponent: {
        component: TestComponent,
      },
    };
    const markdown = render(<Markdown text={text} overrides={overrides} />);

    expect(markdown.find('marquee').length).toEqual(1);
  });

  describe('Allows passing in a custom CodeBlock override', () => {
    it('Accepts a CodeBlock component directly', () => {
      const text = `
# Heading

\`\`\`
var test = true;
\`\`\`
      `;

      const CodeBlock = props => <marquee {...props} />;

      const overrides = {
        CodeBlock: CodeBlock,
      };

      const markdown = mount(<Markdown text={text} overrides={overrides} />);
      expect(markdown.find(CodeBlock).length).toEqual(1);
    });

    it('Accepts a CodeBlock override object including custom props', () => {
      const text = `
# Heading

\`\`\`
var test = true;
\`\`\`
      `;

      const CodeBlock = props => <marquee {...props} />;

      const overrides = {
        CodeBlock: {
          component: CodeBlock,
          props: {
            'data-test': true,
          },
        },
      };

      const markdown = mount(<Markdown text={text} overrides={overrides} />);
      expect(markdown.find(CodeBlock).length).toEqual(1);
      expect(markdown.find(CodeBlock).props()['data-test']).toEqual(true);
    });

    it('When specifying a <code /> element override with a custom CodeBlock override, the CodeBlock wins', () => {
      const text = `
# Heading

\`\`\`js
var test = true;
\`\`\`

\`var test = false;\`
      `;

      const CodeBlock = props => <marquee {...props} />;

      const overrides = {
        CodeBlock,
        code: props => <strong {...props} />,
      };

      const markdown = mount(<Markdown text={text} overrides={overrides} />);

      expect(markdown.find(overrides.CodeBlock).length).toEqual(1);
      // There should only be one <code /> override because the codeblock override overwrote it
      expect(markdown.find(overrides.code).length).toEqual(1);
    });
  });

  it('Renders markdown inside of an html element', () => {
    const markdown = mount(<Markdown text={htmlWrappedMarkdown} />);
    expect(markdown.find('div.test').find('h1').length).toEqual(1);
  });

  it('Renders data attributes on the markdown wrapper', () => {
    const markdown = mount(
      <Markdown text={htmlWrappedMarkdown} data-testid="cool" />
    );
    expect(markdown.find('div[data-testid="cool"]').length).toEqual(1);
  });

  it('Prevents errors on malformed image tags', () => {
    const markdown = mount(
      <Markdown text={`<img src="http://google.com/"></img>`} />
    );
    expect(markdown.find('img').length).toEqual(1);
  });
});
