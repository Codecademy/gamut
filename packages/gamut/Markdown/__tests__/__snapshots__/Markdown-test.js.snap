// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Allows passing in arbitrary react component overrides 1`] = `
<div
  class="spacing-tight"
>
  <h1
    id="heading"
  >
    Heading
  </h1>
  <marquee>
    coooool
  </marquee>
</div>
`;

exports[`Renders html content in markdown 1`] = `
<div
  class="spacing-tight"
>
  <h1>
    Heading 1
  </h1>
  <h3>
    Heading 3
  </h3>
  <table>
    <tbody>
      <tr>
        <td>
          Cool
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="youtubeVideoWrapper"
    data-testid="yt-iframe"
    style="padding-bottom:56.25%"
  >
    <iframe
      allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
      allowfullscreen=""
      frameborder="0"
      src="https://www.youtube.com/embed/KvgrQIK1yPY"
    />
  </div>
</div>
`;

exports[`renders standard Markdown 1`] = `
<div
  class="spacing-tight"
>
  <h1
    id="heading-1"
  >
    Heading 1
  </h1>
  <h3
    id="heading-3"
  >
    Heading 3
  </h3>
  <pre>
    <code
      class="lang-js"
    >
      const jsCode = () =&gt; {
  console.log('hello world);
}
    </code>
  </pre>
</div>
`;
